otter/test/convergence/test_service.py

```diff
@@ -1,95 +1,432 @@
+
+
+class NonConcurrentlyTests(SynchronousTestCase):
+    """Tests for :func:`non_concurrently`."""
+
+    def setUp(self):
+        self.locks = Reference(pset())
+
+    def _get_locks(self):
+        """Get the locks set."""
+        return sync_perform(_get_dispatcher(), self.locks.read())
+
+    def _add_lock(self, value):
+        """Add an item to the locks set."""
+        return sync_perform(_get_dispatcher(),
+                            self.locks.modify(lambda cc: cc.add(value)))
+
+    def test_success(self):
+        """
+        :func:`non_concurrently` returns the result of the passed effect, and
+        adds the ``key`` to the ``locks`` while executing.
+        """
+        dispatcher = _get_dispatcher()
+
+        def execute_stuff():
+            self.assertEqual(self._get_locks(), pset(['the-key']))
+            return 'foo'
+
+        eff = Effect(Func(execute_stuff))
+
+        non_c_eff = non_concurrently(self.locks, 'the-key', eff)
+        self.assertEqual(sync_perform(dispatcher, non_c_eff), 'foo')
+        # after the effect completes, its lock is released
+        self.assertEqual(self._get_locks(), pset([]))
+
+    def test_refuses_concurrency(self):
+        """
+        :func:`non_concurrently` raises :obj:`ConcurrentError` when the key is
+        already locked.
+        """
+        self._add_lock('the-key')
+        eff = Effect(Error(RuntimeError('foo')))
+        non_c_eff = non_concurrently(self.locks, 'the-key', eff)
+        self.assertRaises(
+            ConcurrentError,
+            sync_perform, _get_dispatcher(), non_c_eff)
+        self.assertEqual(self._get_locks(), pset(['the-key']))
+
+    def test_cleans_up_on_exception(self):
+        """
+        When the effect results in error, the key is still removed from the
+        locked set.
+        """
+        dispatcher = _get_dispatcher()
+        eff = Effect(Error(RuntimeError('foo!')))
+        non_c_eff = non_concurrently(self.locks, 'the-key', eff)
+        e = self.assertRaises(RuntimeError, sync_perform, dispatcher,
+                              non_c_eff)
+        self.assertEqual(str(e), 'foo!')
+        self.assertEqual(self._get_locks(), pset([]))
 
 
 class ExecuteConvergenceTests(SynchronousTestCase):
     """Tests for :func:`execute_convergence`."""
 
     def setUp(self):
-        self.state = GroupState('tenant-id', 'group-id', 'group-name',
-                                {}, {}, None, {}, False)
-        self.group = mock_group(self.state, 'tenant-id', 'group-id')
+        self.tenant_id = 'tenant-id'
+        self.group_id = 'group-id'
+        self.state = GroupState(self.tenant_id, self.group_id, 'group-name',
+                                {}, {}, None, {}, False, desired=2)
+        self.group = mock_group(self.state, self.tenant_id, self.group_id)
         self.lc = {'args': {'server': {'name': 'foo'}, 'loadBalancers': []}}
         self.desired_lbs = s(CLBDescription(lb_id='23', port=80))
```

Comment:
Non-blocking thought: I don't like `s` function. It pollutes the namespace with very probable local variable.